### PR TITLE
Consolidate/refactor local exceptions in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR reworks the local exceptions in C#. It should be a model for the other language mappings.

In particular:
 - it reduces the number of local exceptions
 - all local exceptions are now read-only
 - most local exceptions have now no field - only the message and inner exception provided by System.Exception

### CODE REVIEW

File: csharp/src/Ice/LocalExceptions.cs [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+    public ObjectNotExistException()
+    {
+    }
+
+    public ObjectNotExistException(Identity id, string facet, string operation)
+        : base(nameof(ObjectNotExistException), id, facet, operation)
+    {
+    }
+
+    public override string ice_id() => "::Ice::ObjectNotExistException";
```
Comment (bernardnormier) [inline csharp/src/Ice/LocalExceptions.cs:56]: We use `ice_id` for metrics and more.

File: CHANGELOG-3.8.md [hunk 2]
```diff
@@ -58,6 +58,37 @@ These are the changes since the Ice 3.7.10 release in [CHANGELOG-3.7.md](./CHANG
   (the unit is seconds). You can also override this value for a specific object adapter with the configuration
   property `AdapterName.Connection.CloseTimeout`.
 
+- Consolidate and refactor the exceptions derived from LocalException.
+  | Local exception in Ice 3.7          | Replacement                |
+  |-------------------------------------|----------------------------|
+  | BadMagicException                   | ProtocolException (base)   |
```
Comment (bernardnormier) [inline CHANGELOG-3.8.md:64]: Also useful as a guide when updating other language mappings.

File: csharp/src/IceBox/Service.cs [hunk 3]
```diff
@@ -1,50 +1,24 @@
...
 
+#nullable enable
+
 namespace IceBox;
 
 /// <summary>
 /// This exception is a general failure notification.
-/// It is thrown for errors such as a service encountering an error
-///  during initialization, or the service manager being unable to load a service executable.
+/// It is thrown for errors such as a service encountering an error  during initialization, or the service manager
```
Comment (pepone) [inline csharp/src/IceBox/Service.cs:9]:
```suggestion
/// It is thrown for errors such as a service encountering an error during initialization, or the service manager
```

**Review by pepone — APPROVED** [review]

File: csharp/src/Ice/ConnectionI.cs [hunk 4]
```diff
@@ -161,11 +161,18 @@ public void close(ConnectionClose mode)
         {
             if (mode == ConnectionClose.Forcefully)
             {
-                setState(StateClosed, new ConnectionManuallyClosedException(false));
+                setState(StateClosed,
+                    new ConnectionAbortedException(
+                        "Connection close forcefully by the application.",
+                        closedByApplication: true));
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:167]: Not currently - this exception is only thrown when the application aborts the connection "manually".

We need to "closedByApplication" to _not_ retry a connection that was closed manually - that's the existing retry logic.

What about replacing ConnectionIdleException by ConnectionAbortedException with message + closedByApplication = false?

ConnectionLostException (a SocketException) also means "connection aborted". Should we replace it by ConnectionAbortedException? 

All for a follow-up PR.

File: csharp/src/Ice/SSL/TrustManager.cs [hunk 5]
```diff
@@ -43,9 +43,9 @@ internal TrustManager(Ice.Communicator communicator)
                 }
             }
         }
-        catch (RFC2253.ParseException ex)
+        catch (ParseException ex)
         {
-            throw new Ice.PluginInitializationException($"IceSSL: invalid property {key}:\n {ex.reason}");
+            throw new PluginInitializationException($"IceSSL: invalid property {key}", ex);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/SSL/TrustManager.cs:48]: fixed

**Pull request merged**